### DISCUSSION TITLE
getPair call via Ethers.Contract does not return a pair address

### DISCUSSION THREAD

**Discussion by Samboy76** [ConsenSys-archive/truffle#5566] [Truffle Boxes]
Hi,

When I call following **getPair** under uniswapFactory contract I created via ethers.Contract in my javascript it doesn´t return any pair address.

```
const UniswapV2Factory = require("./build/contracts/IUniswapV2Factory.json");

uniswapFactory = new ethers.Contract(
		'0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f', // uniswap v2 factory address in Etherscan mainnet
		UniswapV2Factory.abi, wallet
	);

	const pairAddress = await uniswapFactory.getPair(wethAddress, otherAddress);
  	const pair = new Contract(pairAddress, UniswapV2Pair.abi, wallet);
	console.log("yeah2!");
```
Strange thing is if I removed `await `keywork it returns _Promise { <pending> }_.

However, if I run `await uniswapFactory.getPair(wethAddress, otherAddress);`  directly under Truffle console it resolves the pair address correctly.

Any suggestions would be greatly appreciated as it´s been bugging me also for a very long time.

Thank you

**Comment by apdan510** [comment]
Hey @Samboy76 How do you run your JS script? You mentioned getPair working okay inside Truffle console? Any more screenshots, logs, or info you can share?

Thank you!
-Alex

**Reply by Samboy76** [reply]
Hi @apdan510, at the end I tried also using @uniswap-sdk library so here´s the actual code snippet, which still generates the same issue, along with the output Truffle console commands demonstrating the expected results.

```
const { ChainId, Token, Fetcher, WETH, Pair, TokenAmount, FACTORY_ADDRESS, INIT_CODE_HASH } = require ('@uniswap/sdk');

const provider = new ethers.providers.JsonRpcProvider(process.env.INFURA_URL);
const wallet = new ethers.Wallet(process.env.PRIVATE_KEY, provider);
const otherAddress = '0x111111111117dC0aa78b770fA6A738034120C302'; // Etherscan 1INCH mainnet

const OTHER = new Token(ChainId.MAINNET, otherAddress, 18);
const uniswapPair = await Fetcher.fetchPairData(OTHER, WETH[OTHER.chainId], provider,);
```

**Output Truffle console commands**

**truffle(mainnet_fork)> const { INFURA_URL, METAMASK_ADDRESS } = process.env;**
undefined
**truffle(mainnet_fork)> const { ChainId, Token, Fetcher, WETH, Pair, TokenAmount, FACTORY_ADDRESS, INIT_CODE_HASH } = require ('@uniswap/sdk');**
undefined
**truffle(mainnet_fork)> const provider = new ethers.providers.JsonRpcProvider(process.env.INFURA_URL);**
undefined
**truffle(mainnet_fork)> const wallet = new ethers.Wallet(process.env.PRIVATE_KEY, provider);**
undefined
**truffle(mainnet_fork)> const otherAddress = '0x111111111117dC0aa78b770fA6A738034120C302';**
undefined
**truffle(mainnet_fork)> const OTHER = new Token(ChainId.MAINNET, otherAddress, 18);**
undefined
**truffle(mainnet_fork)> const uniswapPair = await Fetcher.fetchPairData(OTHER, WETH[OTHER.chainId], provider,);**
undefined
**truffle(mainnet_fork)> console.log("uniswapPair ", uniswapPair);**
_uniswapPair  Pair {
  liquidityToken: Token {
    decimals: 18,
    symbol: 'UNI-V2',
    name: 'Uniswap V2',
    chainId: 1,
    address: '0x26aAd2da94C59524ac0D93F6D6Cbf9071d7086f2'
  },
  tokenAmounts: [
    TokenAmount {
      numerator: [JSBI],
      denominator: [JSBI],
      currency: [Token],
      token: [Token]
    },
    TokenAmount {
      numerator: [JSBI],
      denominator: [JSBI],
      currency: [Token],
      token: [Token]
    }
  ]
}_
undefined
**truffle(mainnet_fork)> console.log("uniswapPair.reserve0 ", uniswapPair.reserve0);**
_uniswapPair.reserve0  TokenAmount {
  numerator: JSBI(3) [ 909525627, 1065416368, 151027, sign: false ],
  denominator: JSBI(2) [ 660865024, 931322574, sign: false ],
  currency: Token {
    decimals: 18,
    symbol: undefined,
    name: undefined,
    chainId: 1,
    address: '0x111111111117dC0aa78b770fA6A738034120C302'
  },
  token: Token {
    decimals: 18,
    symbol: undefined,
    name: undefined,
    chainId: 1,
    address: '0x111111111117dC0aa78b770fA6A738034120C302'
  }
}_
undefined
**truffle(mainnet_fork)> console.log("uniswapPair.reserve1 ", uniswapPair.reserve1);**
_uniswapPair.reserve1  TokenAmount {
  numerator: JSBI(3) [ 1072296449, 41327450, 69, sign: false ],
  denominator: JSBI(2) [ 660865024, 931322574, sign: false ],
  currency: Token {
    decimals: 18,
    symbol: 'WETH',
    name: 'Wrapped Ether',
    chainId: 1,
    address: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2'
  },
  token: Token {
    decimals: 18,
    symbol: 'WETH',
    name: 'Wrapped Ether',
    chainId: 1,
    address: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2'
  }
}_
undefined

**truffle(mainnet_fork)> truffle version**
Truffle v5.5.31 (core: 5.5.31)
Ganache v7.4.3
Solidity - >=0.5.0 <0.9.0 (solc-js)
Node v16.17.0
Web3.js v1.7.4

Appreciate you help.

Thank you very much!

**Reply by Samboy76** [reply]
Also need help deciphering how to extract the reserve amounts from `uniswapPair.reserve0` and `uniswapPair.reserve1` commands as can only see numerator and denominator values inside them.

Thank you

**Reply by Samboy76** [reply]
INFURA_URL = “ https://mainnet.infura.io/v3/“ and private API key together

**Reply by Samboy76** [reply]
Hi @apdan510 any luck? FYI I have tried both methods via uniswap v2 sdk method above and through ethers.Contract as shown in below code snippet

```
// SDK´s off-chain method of getPair function
	const uniswapEthOtherAddr = getCreate2Address(
		uniswapFactory.address,
		keccak256(['bytes'], [pack(['address', 'address'], [otherAddress, wethAddress])]),
		INIT_CODE_HASH
	);
	uniswapEthOther = new ethers.Contract(
		uniswapEthOtherAddr,
		UniswapV2Pair.abi, 
		provider
	);

        const reserves = await uniswapEthOther.functions.getReserves();
```

Neither code method works. If I remove await keyword in both it always returns with Promise { <pending> }. If I put it back in, it doesn´t print any output or get resolved. It just calls the callback() function and halts JS execution.

I´m desperate as it seems under Truffle runtime, it fails on making certain on-chain ethers.Contract calls to certain functions such as getReserves() which I need to perform for my script.

Thanks for your help. Much appreciated!

**Reply by apdan510** [reply]
let me dig into your code a bit. Thanks for the info!

**Reply by apdan510** [reply]
@Samboy76 could you please open a Support ticket with us for better 1 on 1 support? It will be more discreet. The team may need to test and try to reproduce your issues with your complete repo if that is okay with you. I will look out for your ticket to make a priority. There is something strange with the JS script. The await for a promise is correct, but weird that it is returning no data. 

https://trufflesuite.zendesk.com/hc/en-us/requests/new

**Reply by Samboy76** [reply]
Have raised Truffle Ticket #**945282**. Also shared the repo pointing to priceBotChecker.js which is the issue.
Much appreciate your help!

**Reply by apdan510** [reply]
Saw the ticket. Thank you! Ill get more eyes on this issue while I am sifting through the code. Thank you!

**Reply by Samboy76** [reply]
Thanks a lot @apdan510.
Looking fwd to your finds...

**Reply by Samboy76** [reply]
https://github.com/Samboy76/Flashloan-arbitrage

**Reply by Samboy76** [reply]
priceBotChecker.js

**Comment by apdan510** [comment]
Got the repo. Thank you! Will hop on analysis with the rest
of team tomorrow morning. Hoping its a small bug somewhere.

**Reply by apdan510** [reply]
* This is a suspected bug that happens in a Truffle project when working with Uniswap pair addresses. There are no errors when this project is converted to a HH project *

**Reply by Samboy76** [reply]
It goes well beyond in that the issue lies in its inability to make any on-chain call outs to the network not just limited to Uniswap pair addresses. This was the case for me because I couldn´t even retrieve the pair´s reserves through getReserves() even when I attempted to retrieve the pairAddress off-chain like through **CREATE2** beforehand as shown on the [docs](https://docs.uniswap.org/sdk/2.0.0/guides/getting-pair-addresses).
